### ISSUE TITLE
getdata() method no longer works with tile-compressed uint images

### ISSUE THREAD

**Issue by rsiverd** [astropy/astropy#4588]
Recently (by v 1.1.1), the getdata() convenience method has started failing on tile-compressed USHORT images (BITPIX 16). Floating-point images appear to still work fine. As recently as astropy v 1.0.3, the getdata() method worked just fine with compressed and uncompressed images regardless of data type. 

The error I get is:

```
TypeError: Cannot cast ufunc add output from dtype('float64') to dtype('uint16') with casting rule 'same_kind'
```

For reference, I'm running Python 2.7.11 with numpy 1.10.2 [MKL accelerated] from anaconda on CentOS 6 64-bit.

A few permutations of this command fail. I can avoid the error by specifying do_not_scale_image_values=True, but this changes the values reported (as expected).

```
import astropy.io.fits as pf
idata = pf.getdata('image.fits.fz')   # fails with TypeError
idata = pf.getdata('image.fits.fz', uint=True, do_not_scale_image_values)   # also fails
idata = pf.getdata('image.fits.fz', uint=True, do_not_scale_image_data=True)  # no error (but wrong values as expected)
```

Any ideas?

[getdata_error.txt](https://github.com/astropy/astropy/files/122652/getdata_error.txt)

**Comment by marcopigg** [comment]
I have attempted a fix to this issue in the PR above. However, I still have problems with big (>250MB), multi-extension FITS files, for which I get the following exception:
```
Exception AttributeError: AttributeError('_coldefs',) in <bound method FITS_rec.__del__ of FITS_rec([(array([1042,    0], dtype=int32),),
       (array([1043, 1042], dtype=int32),),
       (array([1039, 2085], dtype=int32),), ...,
       (array([   2141, 8978460], dtype=int32),),
       (array([   2173, 8980601], dtype=int32),),
       (array([   2120, 8982774], dtype=int32),)], 
      dtype=(numpy.record, [('COMPRESSED_DATA', '>i4', (2,))]))> ignored
```

**Comment by saimn** [comment]
This was likely introduced by Numpy 1.10 (*Default casting for inplace operations will change to ‘same_kind’ in Numpy 1.10.0* http://docs.scipy.org/doc/numpy-1.10.1/release.html#future-changes).

**Comment by marcopigg** [comment]
I am getting the same error (for example, when running the astrometry.net code) on unsigned integer (uncompressed) data.
I tried patching the `scale` function in `image.py` in the same way as `compressed.py`, i.e.
```
442           -self.data += -_zero
442           +self.data += -self.data.dtype.type(_zero)
```
but I get a test fail.

**Comment by astroleo** [comment]
I get a similar (but not the same error) when trying to access scaled integer values in an image cube within a binary table. I have put an example file to ftp://ftp.mpe.mpg.de/people/burtscher/MIDI.2013-05-03T01_00_59.000.fits (17.4 MB).

I open the file with
```
from astropy.io import fits
hdu=fits.open(file)
```

hdu.info() looks like this:
```
...
7    IMAGING_DATA  BinTableHDU    107   1000R x 17C   ['1J', '1D', '1E', '2I', '2I', '1I', '2D', '2D', '2E', '1E', '1I', '1I', '2A', '1I', '2A', '4278I', '4278I']
...

```
And hdu[7].data.columns looks like this:

```
ColDefs(
    [...other tables...]
    name = 'DATA1'; format = '4278I'; unit = 'ADU'; bscale = 1.0; bzero = 32768.0; dim = '(62,69)'
    name = 'DATA2'; format = '4278I'; unit = 'ADU'; bscale = 1.0; bzero = 32768.0; dim = '(62,69)'
)

```
When trying to access these image cubes I get the following error:


```
In [32]: img=hdu[7].data["DATA1"]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-32-426539ce5f7e> in <module>()
----> 1 img=hdu[7].data["DATA1"]

/usr/local/lib/python3.5/site-packages/astropy/io/fits/fitsrec.py in __getitem__(self, key)
    482     def __getitem__(self, key):
    483         if isinstance(key, string_types):
--> 484             return self.field(key)
    485         elif isinstance(key, (slice, np.ndarray, tuple, list)):
    486             # Have to view as a recarray then back as a FITS_rec, otherwise the

/usr/local/lib/python3.5/site-packages/astropy/io/fits/fitsrec.py in field(self, key)
    656                 # Handle all other column data types which are fixed-width
    657                 # fields
--> 658                 converted = self._convert_other(column, field, recformat)
    659
    660             self._converted[name] = converted

/usr/local/lib/python3.5/site-packages/astropy/io/fits/fitsrec.py in _convert_other(self, column, field, recformat)
    875                         field = test_overflow
    876                 else:
--> 877                     field += bzero
    878         elif _bool and field.dtype != bool:
    879             field = np.equal(field, ord('T'))

TypeError: Cannot cast ufunc add output from dtype('float64') to dtype('uint16') with casting rule 'same_kind'
```

The workaround mentioned above, i.e. using the getdata convenience method with the options uint=False and do_not_scale_image_data=False does not work for me. The only way to read the data at the moment is to remove the header cards describing the scaling, i.e.

```
hdu[7].header.remove("TSCAL16")
hdu[7].header.remove("TSCAL17")
hdu[7].header.remove("TZERO16")
hdu[7].header.remove("TZERO17")
```

and only then accessing the data (which will then of course be unscaled).

**Comment by saimn** [comment]
I think this was fixed by #5527
